src/workerd/api/eventsource.c++

```diff
@@ -0,0 +1,526 @@
+    self->start(js);
+  }), JSG_VISITABLE_LAMBDA((self=JSG_THIS),(self),(jsg::Lock& js, jsg::Value exception) {
+    // In this case, it is most likely the EventSource was closed by the user or
+    // there was some other failure. We should not continue trying to reconnect.
+    self->notifyError(js, jsg::JsValue(exception.getHandle(js)));
+  }));
+}
+
+void EventSource::start(jsg::Lock& js) {
+  auto& i = KJ_ASSERT_NONNULL(impl);
+  if (readyState == State::CLOSED) return;
+
+  auto fetcher = i.options.fetcher.map([](jsg::Ref<Fetcher>& f) { return f.addRef(); });
+
+  auto onSuccess = JSG_VISITABLE_LAMBDA(
+      (self=JSG_THIS,
+       fetcher = fetcher.map([](jsg::Ref<Fetcher>& f) -> jsg::Ref<Fetcher> { return f.addRef(); })),
+      (self, fetcher),
+      (jsg::Lock& js, jsg::Ref<Response> response) {
+    if (self->readyState == State::CLOSED) return js.resolvedPromise();
+    auto& impl = KJ_ASSERT_NONNULL(self->impl);
+    if (!response->getOk()) {
+      // Response status code is not 2xx, so we fail.
+      // No reconnection attempt should be made.
+      auto message = kj::str("The response status code was ", response->getStatus());
+      auto ex = js.domException(kj::str("AbortError"), kj::mv(message));
+      auto handle = KJ_ASSERT_NONNULL(ex.tryGetHandle(js));
+      self->notifyError(js, jsg::JsValue(handle));
+      return js.resolvedPromise();
+    }
+
+    // TODO(cleanup): Using jsg::ByteString here is really annoying. It would be nice to have
+    // an internal alternative that doesn't require an allocation.
+    KJ_IF_SOME(contentType, response->getHeaders(js)->get(
+        jsg::ByteString(kj::str("content-type")))) {
+      bool invalid = false;
+      KJ_IF_SOME(parsed, MimeType::tryParse(contentType)) {
+        invalid = parsed != MimeType::EVENT_STREAM;
+      } else {
+        invalid = true;
+      }
+      if (invalid) {
+        // No reconnection attempt should be made.
+        auto message = kj::str("The content type '", contentType, "' is invalid.");
+        auto ex = js.domException(kj::str("AbortError"), kj::mv(message));
+        auto handle = KJ_ASSERT_NONNULL(ex.tryGetHandle(js));
+        self->notifyError(js, jsg::JsValue(handle));
+        return js.resolvedPromise();
+      }
+    } else {
+      // No reconnection attempt should be made.
+      auto message = kj::str("No content type header was present in the response.");
+      auto ex = js.domException(kj::str("AbortError"), kj::mv(message));
+      auto handle = KJ_ASSERT_NONNULL(ex.tryGetHandle(js));
+      self->notifyError(js, jsg::JsValue(handle));
+      return js.resolvedPromise();
+    }
+
+    // If the request was redirected, update the URL to the new location.
+    if (response->getRedirected()) {
+      KJ_IF_SOME(newUrl, jsg::Url::tryParse(response->getUrl())) {
+        impl.url = kj::mv(newUrl);
+      } else {}  // Extra else block to squash compiler warning
```

Comment:
In this case there's likely always a URL. This could throw an error but at this point there's no harm in just skipping if there's no url. (the `Response` object will always have a URL).